WooCommerce/Classes/ViewRelated/Editor/AztecEditorViewController.swift

```diff
@@ -0,0 +1,863 @@
+        }
+        // Convert the keyboard frame from window base coordinate
+        currentKeyboardFrame = view.convert(keyboardFrame, from: nil)
+        refreshInsets(forKeyboardFrame: keyboardFrame)
+    }
+
+    @objc func keyboardDidHide(_ notification: Foundation.Notification) {
+        guard
+            let userInfo = notification.userInfo as? [String: AnyObject],
+            let keyboardFrame = (userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue
+            else {
+                return
+        }
+
+        currentKeyboardFrame = .zero
+        refreshInsets(forKeyboardFrame: keyboardFrame)
+    }
+
+    func refreshInsets(forKeyboardFrame keyboardFrame: CGRect) {
+        let referenceView = editorView.activeView
+
+        let contentInsets  = UIEdgeInsets(top: referenceView.contentInset.top, left: 0, bottom: view.frame.maxY - (keyboardFrame.minY + self.view.layoutMargins.bottom), right: 0)
+
+        htmlTextView.contentInset = contentInsets
+        richTextView.contentInset = contentInsets
+
+        updateScrollInsets()
+    }
+
+    func updateScrollInsets() {
+        let referenceView = editorView.activeView
+        var scrollInsets = referenceView.contentInset
+        var rightMargin = (view.frame.maxX - referenceView.frame.maxX)
+        rightMargin -= view.safeAreaInsets.right
+        scrollInsets.right = -rightMargin
+        referenceView.scrollIndicatorInsets = scrollInsets
+    }
+}
+
+// MARK: - Notifications
+//
+private extension AztecEditorViewController {
+    func startListeningToNotifications() {
+        let nc = NotificationCenter.default
+        nc.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
+        nc.addObserver(self, selector: #selector(keyboardDidHide), name: UIResponder.keyboardDidHideNotification, object: nil)
+        nc.addObserver(self, selector: #selector(applicationWillResignActive(_:)), name: UIApplication.willResignActiveNotification, object: nil)
+    }
+
+    /// Unregisters from the Notification Center
+    ///
+    func stopListeningToNotifications() {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    @objc func applicationWillResignActive(_ notification: Foundation.Notification) {
+
+        // [2018-03-05] Need to close the options VC on backgrounding to prevent view hierarchy inconsistency crasher.
+        optionsTablePresenter.dismiss()
+    }
+}
+
+// MARK: - Navigation actions
+//
+private extension AztecEditorViewController {
+    @objc func cancelButtonTapped() {
+        navigationController?.popViewController(animated: true)
+    }
+
+    @objc func saveButtonTapped() {
+        let content = getHTML()
+        onContentSave?(content)
+
+        navigationController?.popViewController(animated: true)
+    }
+}
+
+// MARK: - TextViewAttachmentDelegate Conformance
+//
+extension AztecEditorViewController: TextViewAttachmentDelegate {
```

Comment:
And this implementation of the TextViewAttachmentDelegate, as it is now, could also be extracted.